// @waves/waves-api/raw/src/tools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var waves_signature_generator_1 = require("@waves/waves-signature-generator");
var constants_1 = require("./constants");
var base58_1 = require("./libs/base58");
var crypto_1 = require("./utils/crypto");
var transactions_1 = require("./utils/transactions"); // TODO : fix this issue with interface
exports.default = {
    getAddressFromPublicKey: function (publicKey) {
        var publicKeyBytes = base58_1.default.decode(publicKey);
        return crypto_1.default.buildRawAddress(publicKeyBytes);
    },
    calculateTimeDiff: function (nodeTime, userTime) {
        return nodeTime - userTime;
    },
    base58: {
        encode: base58_1.default.encode,
        decode: base58_1.default.decode
    },
    getMinimumDataTxFee: function (data) {
        var emptyDataTx = new waves_signature_generator_1.TX_TYPE_MAP.data({
            senderPublicKey: '11111111111111111111111111111111',
            timestamp: 0,
            fee: '',
            data: data
        });
        return emptyDataTx.getBytes().then(function (bytes) { return Math.ceil(bytes.length / 1024) * constants_1.MINIMUM_DATA_FEE_PER_KB; });
    },
    createTransaction: transactions_1.createTransaction
};
